# Question 1: Did Quadeca's Musical Themes and Musical Style Change From the Period Before Insecure Compared to the Period After Insecure?

---

## Setup

In [1]:
import os
import re
import csv
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from nltk import tokenize
from nltk.sentiment.vader import SentimentIntensityAnalyzer

In [2]:
# import other notebooks
%run /Commjhub/jupyterhub/comm318_fall2019/raymon-shi/comm313_S21/comm313_S21_Final_Project/data/functions.ipynb

In [3]:
# import global variables
%store -r b_album_one_lyrics
%store -r b_album_two_lyrics
%store -r b_album_three_lyrics
%store -r b_album_four_lyrics
%store -r a_album_one_lyrics
%store -r a_album_two_lyrics

sid = SentimentIntensityAnalyzer()

---------------------------------------

## Sentiment Analysis of Keywords from Keyness Analysis

### Keywords of Albums Before the Insecure Diss Track

In [4]:
# keywords from the before album corpus
b_corpus_kw = ['and', 'people', 'im', 'man', 'is', 'know', 'am', 'that', 'really', "i've", 'find', 'livin', 'feeling',
              'it', 'we', 'ever', 'to', 'game', 'thought', 'of', 'about', 'will']

In [38]:
print('the before corpus keywords lexicon score is:', get_kw_vader_score(b_corpus_kw))

the before corpus keywords lexicon score is: 0.022727272727272728


### Observations of Keywords Sentiment of all Albums Before the Insecure Diss Track

According to the vader polarity score, it seems that the keywords from the corpus of the albums before the Insecure Diss Track seem to have a very neutral sentiment.

### Keywords of Albums After the Insecure Diss Track

In [39]:
a_corpus_kw = ['uh', 'yeah', 'ooh', 'work', 'gon', 'until', 'still', 'real', 'nothing', 'did', 'was', 'they', 'had',
              'everything', 'ayy', 'fuckin', 'he', "didn't", 'nothin', 'long', 'bitch', 'think', 'i', 'alone', 'left',
              'she', 'myself', 'matter', 'old', 'if', 'only', 'used', 'looking', 'why', 'wrong', 'love', 'thank', 'were',
              "can't", 'em', 'back', 'turn', 'wish', 'anymore', 'sorry']

In [40]:
print('the after corpus keywords lexicon score is:', get_kw_vader_score(a_corpus_kw))

the after corpus keywords lexicon score is: 0.03333333333333334


### Observations of Keyword Sentiment of all Albums After the Insecure Diss Track

According to the vader lexicon score, it seems that the keywords from the corpus of the albums after the Insecure Diss Track also seem to have a very neutral sentiment.

### Overall Observation of Keyword Sentiment from Before and After Albums

Overall, it seems that the keywords that were more likely to appear in one corpus compared to the other were keywords that had mostly neutral sentiment. It appears that trying to use the keywords to predict the sentiment of the corpus does not help. The next step will be to do a sentiment analysis on each song to see how many positive, negative, and neutral songs exists within each album. From there, I will do a sentiment analysis over the albums, to get the overall sentiment score of each album in each period of time and then compare the two periods of time. Afterward, I will see if the sentiment scores supports the other analyses that I've conducted.

---------------------------------------

## Sentiment Analysis of Songs

### Albums Before the Insecure Diss Track

#### Work in Progress (2015)

In [10]:
print('the vader polarity average scores for each song in the album are:')
b_aone_polarity = print_polarity_score_album(b_album_one_lyrics)
b_aone_polarity

the vader polarity average scores for each song in the album are:


[0.9949,
 0.9883,
 -0.9593,
 0.9649,
 0.9976,
 -0.971,
 0.7287,
 -0.9658,
 0.1177,
 -0.535,
 -0.1952,
 -0.9985,
 0.8782,
 -0.9907,
 0.9059]

In [11]:
b_aone_pos_songs = list()
b_aone_neg_songs = list()
b_aone_neu_songs = list()

sort_sentiment_songs(b_aone_polarity, b_aone_pos_songs, b_aone_neg_songs, b_aone_neu_songs)

There are 8 positive songs in this album
There are 7 negative songs in this album
There are 0 neutral songs in this album


#### Nostalgia for the Now (2016)

In [13]:
print('the vader polarity average scores for each song in the album are:')
b_atwo_polarity = print_polarity_score_album(b_album_two_lyrics)
b_atwo_polarity

the vader polarity average scores for each song in the album are:


[-0.384,
 -0.9927,
 0.9972,
 0.9976,
 -0.9927,
 0.7397,
 0.9953,
 0.9519,
 -0.8424,
 -0.8982,
 -0.9946,
 -0.489,
 -0.9979,
 0.8911,
 0.8995]

In [14]:
b_atwo_pos_songs = list()
b_atwo_neg_songs = list()
b_atwo_neu_songs = list()

sort_sentiment_songs(b_atwo_polarity, b_atwo_pos_songs, b_atwo_neg_songs, b_atwo_neu_songs)

There are 7 positive songs in this album
There are 8 negative songs in this album
There are 0 neutral songs in this album


#### Bad Internet Rapper (2017)

In [16]:
print('the vader polarity average scores for each song in the album are:')
b_athree_polarity = print_polarity_score_album(b_album_three_lyrics)
b_athree_polarity

the vader polarity average scores for each song in the album are:


[0.9866,
 0.9894,
 0.9958,
 0.9935,
 -0.995,
 -0.8292,
 0.5963,
 0.9959,
 0.9969,
 0.9934,
 0.9645,
 -0.9411]

In [17]:
b_athree_pos_songs = list()
b_athree_neg_songs = list()
b_athree_neu_songs = list()

sort_sentiment_songs(b_athree_polarity, b_athree_pos_songs, b_athree_neg_songs, b_athree_neu_songs)

There are 9 positive songs in this album
There are 3 negative songs in this album
There are 0 neutral songs in this album


#### Out of Order (2017)

In [19]:
print('the vader polarity average scores for each song in the album are:')
b_afour_polarity = print_polarity_score_album(b_album_four_lyrics)
b_afour_polarity

the vader polarity average scores for each song in the album are:


[-0.9603,
 0.8933,
 -0.8871,
 0.9954,
 -0.9826,
 -0.9413,
 0.9925,
 -0.9994,
 -0.9341,
 -0.8602,
 0.757,
 -0.9666]

In [20]:
b_afour_pos_songs = list()
b_afour_neg_songs = list()
b_afour_neu_songs = list()

sort_sentiment_songs(b_afour_polarity, b_afour_pos_songs, b_afour_neg_songs, b_afour_neu_songs)

There are 4 positive songs in this album
There are 8 negative songs in this album
There are 0 neutral songs in this album


#### All Albums Before the Insecure Diss Track

In [21]:
b_pos_songs = len(b_aone_pos_songs) + len(b_atwo_pos_songs) + len(b_athree_pos_songs) + len(b_afour_pos_songs)
b_neg_songs = len(b_aone_neg_songs) + len(b_atwo_neg_songs) + len(b_athree_neg_songs) + len(b_afour_neg_songs)
b_neu_songs = len(b_aone_neu_songs) + len(b_atwo_neu_songs) + len(b_athree_neu_songs) + len(b_afour_neu_songs)

print('There are a total of', b_pos_songs, 'positive songs within this time period')
print('There are a total of', b_neg_songs, 'negative songs within this time period')
print('There are a total of', b_neu_songs, 'neutral songs within this time period')

There are a total of 28 positive songs within this time period
There are a total of 26 negative songs within this time period
There are a total of 0 neutral songs within this time period


#### Overall Observations of Songs from all Albums Before the Insecure Diss Track

According to the vader polarity scores, there are a total of 28 positive songs, 26 negative songs, and 0 neutral songs. From this information, I would assume that the overall sentiment from the albums before the Insecure diss track would be neutral to slightly more positive due to there being more positive songs than there are negative songs.

### Albums After the Insecure Diss Track

#### Voice Memos (2019)

In [23]:
print('the vader polarity average scores for each song in the album are:')
a_aone_polarity = print_polarity_score_album(a_album_one_lyrics)
a_aone_polarity

the vader polarity average scores for each song in the album are:


[0.9881,
 0.9975,
 0.9827,
 0.9729,
 0.9967,
 0.9905,
 0.9973,
 -0.658,
 -0.9987,
 0.9926,
 -0.3506,
 -0.8705,
 0.9902]

In [24]:
a_aone_pos_songs = list()
a_aone_neg_songs = list()
a_aone_neu_songs = list()

sort_sentiment_songs(a_aone_polarity, a_aone_pos_songs, a_aone_neg_songs, a_aone_neu_songs)

There are 9 positive songs in this album
There are 4 negative songs in this album
There are 0 neutral songs in this album


#### From Me To You (2021)

In [26]:
print('the vader polarity average scores for each song in the album are:')
a_atwo_polarity = print_polarity_score_album(a_album_two_lyrics)
a_atwo_polarity

the vader polarity average scores for each song in the album are:


[0.9091,
 0.6872,
 0.9664,
 -0.9801,
 0.9979,
 -0.5204,
 0.9768,
 -0.9829,
 -0.9878,
 -0.8512,
 0.4019,
 -0.942,
 0.9166,
 0.4153,
 -0.9688,
 -0.9463,
 0.9883,
 0.1027]

In [27]:
a_atwo_pos_songs = list()
a_atwo_neg_songs = list()
a_atwo_neu_songs = list()

sort_sentiment_songs(a_atwo_polarity, a_atwo_pos_songs, a_atwo_neg_songs, a_atwo_neu_songs)

There are 10 positive songs in this album
There are 8 negative songs in this album
There are 0 neutral songs in this album


#### All Albums After the Insecure Diss Track

In [28]:
a_pos_songs = len(a_aone_pos_songs) + len(a_atwo_pos_songs)
a_neg_songs = len(a_aone_neg_songs) + len(a_atwo_neg_songs)
a_neu_songs = len(a_aone_neu_songs) + len(a_atwo_neu_songs)

print('There are a total of', a_pos_songs, 'positive songs within this time period')
print('There are a total of', a_neg_songs, 'negative songs within this time period')
print('There are a total of', a_neu_songs, 'neutral songs within this time period')

There are a total of 19 positive songs within this time period
There are a total of 12 negative songs within this time period
There are a total of 0 neutral songs within this time period


#### Overall Observations of Songs from all Albums After the Insecure Diss Track

According to the vader polarity scores, there are a total of 19 positive songs, 12 negative songs, and 0 neutral songs. From this information, I would assume that the overall sentiment from the albums after the Insecure diss track would be more positive because there are more positive songs than there are negative songs in the after period.

### Overall Observation between the Before and After Albums

After reviewing the information of the total amount of positive, negative, and neutral songs, I'm surprised to see that there are more positive songs than there are negative songs. In the before corpus, there are a total of 51.85% positive songs and 48.15% negative songs. In the after corpus, there are 61.29% positive songs and 38.71% negative songs. Just judging from the percentages, there seems to be an increase in positive songs and decrease in negative songs in the after corpus compared to the before corpus. The next step will be to investigate the average sentiment for each album and then figure out the average sentiment for the time period and see if the same results are found.

---

## Sentiment Analysis of Albums

### Albums Before the Insecure Diss Track

#### Work in Progress (2015)

In [29]:
vps_b_one_score = get_polarity_score(b_album_one_lyrics)

print('the vader polarity sccore is:', vps_b_one_score)

the vader lexicon score is: 0.0647143438959433
the vader polarity sccore is: 0.06404666666666661


#### Nostalgia for the Now (2016)

In [30]:
vps_b_two_score = get_polarity_score(b_album_two_lyrics)

print('the vader polarity sccore is:', vps_b_two_score)

the vader lexicon score is: 0.059160983310107725
the vader polarity sccore is: -0.00794666666666671


#### Bad Internet Rapper (2017)

In [31]:
vps_b_three_score = get_polarity_score(b_album_three_lyrics)

print('the vader polarity sccore is:', vps_b_three_score)

the vader lexicon score is: 0.06488476964985307
the vader polarity sccore is: 0.47891666666666666


#### Out of Order (2017)

In [32]:
vps_b_four_score = get_polarity_score(b_album_four_lyrics)

print('the vader polarity sccore is:', vps_b_four_score)

the vader lexicon score is: 0.06776892386306003
the vader polarity sccore is: -0.32444999999999996


#### Overall Sentiment Score for all Albums Before the Insecure Diss Track

In [33]:
vlp_b_album_score = (vps_b_one_score + vps_b_two_score + vps_b_three_score + vps_b_four_score) / 4

print('the overall vader polarity score is:', vlp_b_album_score)

the overall vader lexicon score is: 0.06413225517974103
the overall vader polarity score is: 0.052641666666666656


#### Overall Observations for all Albums Before the Insecure Diss Track

According to the overall average vader polarity score, the albums before the Insecure diss track seem to have a vader polarity score of 0.053. Since this value is greater than or equal to 0.05, this means that the overall corpus has a positive sentiment. Therefore, the albums in the before period are generally positive sentiment albums.

### Albums After the Insecure Diss Track

#### Voice Memos (2019)

In [34]:
vps_a_one_score = get_polarity_score(a_album_one_lyrics)

print('the vader polarity sccore is:', vps_a_one_score)

the vader lexicon score is: 0.060764818861591394
the vader polarity sccore is: 0.4639


#### From Me To You (2021)

In [35]:
vps_a_two_score = get_polarity_score(a_album_two_lyrics)

print('the vader polarity sccore is:', vps_a_two_score)

the vader lexicon score is: 0.06466207074089213
the vader polarity sccore is: 0.01014999999999998


#### Overall Polarity Score for all Albums Before the Insecure Diss Track

In [36]:
vlp_a_album_score = (vps_a_one_score + vps_a_two_score) / 2

print('the overall vader polarity sccore is:', vlp_a_album_score)

the overall vader lexicon score is: 0.06271344480124176
the overall vader polarity sccore is: 0.23702499999999999


In [37]:
sid.polarity_scores(a_album_one_lyrics[4])

{'compound': 0.9967, 'neg': 0.103, 'neu': 0.722, 'pos': 0.174}

In [44]:
a_album_one_lyrics[4]

'And when you may try\nTo hide away\nWhy do you hide away? (Ooh, ooh, ooh)\nWhy do you hide away? (Ooh, ooh, ooh)\nFeeling like there’s no room left for me (Ooh, ooh, ooh)\nFeeling like there’s (Ooh, ooh, ooh)\nNothing left (Ooh, ooh, ooh, ooh)\nAnd there’s nothing more (Ooh, ooh, ooh)\n\n\nUsed to give a fuck, that was just a old phase (Old phase)\nYou and me were close, now you further than the old days (Old days)\nSometimes, I wonder why I’m headed towards a cold place (Yeah, yeah, yeah, yeah)\nThey like, “Let me fix you”, I don’t fuck with Coldplay (No)\nYeah, wish I could forget, I cannot forgive, yeah\nEvery step- I gotta relive, yeah\nTake a breath, feel it in my ribs, yeah\nLet my voice fade into the mist\nMist, mist, mist...\n\nYeah!\nI been talking to the man sitting on my left shoulder (Shoulder)\nHe said "Everything just gets worse when you\'re older" ()\nI just pulled up to the function, can’t even function (No, no)\nKnow nothin\', I don’t know nothin\' (No, no)\n\nI’m not

#### Overall Observations for all Albums After the Insecure Diss Track

According to the overall average vader polarity score, the albums after the Insecure diss track seem to have a vader score of 0.237. Since this value is greater than or equal to 0.05, this means that the overall corpus has a positive sentiment. Therefore, the albums in the after period are generally positive sentiment albums.

### Overall Observations of Sentiment Analysis of Album Lyrics from Before and After

After calculating the average polarity score for each time period, I was once again surprised to find that both time periods produced generally positive sentiment albums. The average polarity score for an album in the before period was 0.053 and the average polarity score for an album in the after period was 0.237. 

This result is significant for two main reasons:
> In the earlier analysis, I found that there were some common themes that were talked about in both corpuses, such as struggling with depression, dealing with public expectations, and talking about past romantic relationships. As a result, I thought that the sentiment scores between the two corpuses would be more or less the same. However, there is a significant difference between the sentiment score of the before corpus and the sentiment score of the after corpus.

>From those themes that I discovered that Quadeca was rapping about, I expected more albums to have an overall negative sentiment and that both corpuses would have an overall negative sentiment value. However, it appears that both corpuses had positive overall sentiment values. This could mean that Quadeca is talking about some of these more serious and negative topics and the vader polarity score is misinterpreting/overcorrecting the sentiment of the song. An example of this can be seen in the the after album Voice Memos (2019) song titled "The Man on My Left Shoulder (a_album_one_lyrics[4])." This song is talking about Quadeca and how he's dealing with an inner voice that always points out the negativity in life. This song is definitely a negative sentiment song, yet the vader compound polarity score returns 0.9967, which makes it seem like an extremely happy song when it's not. Another possibility for this result is that the keyness analysis and the keyword in context concordance analysis that I did overestimated the themes that were suppose to be in the albums.
